### PR TITLE
Do not update FCT masks between RK stages

### PR DESCRIPTION
**Pull request by trhille**
Do not update FCT masks between RK stages. The masking for FCT is now calculated in the first RK stage and not updated again in the time step.

### CODE REVIEW

**Review by mperego — APPROVED** [review]
It looks good to me. Thanks.

File: components/mpas-albany-landice/src/mode_forward/mpas_li_time_integration_fe_rk.F [hunk 1]
```diff
@@ -279,6 +282,17 @@ subroutine li_time_integrator_forwardeuler_rungekutta(domain, err)
...
       deltatFull = deltat ! Save deltat in order to reset it at end of RK loop
 
+      if ( (trim(config_tracer_advection) == 'fct') .or. &
+           (trim(config_thickness_advection) == 'fct') ) then
+         call li_tracer_advection_fct_shared_init(geometryPool, err_tmp)
+         if (err_tmp /= 0) then
+            err = 1
+            call mpas_log_write(                                 &
+               'Error encountered during fct tracer advection shared init', &
+               MPAS_LOG_ERR, masterOnly=.true.)
```
Comment (matthewhoffman) [inline components/mpas-albany-landice/src/mode_forward/mpas_li_time_integration_fe_rk.F:292]: Why `masterOnly`?  It seems possible that the error only occurs on a different processor and then it doesn't get written anywhere.  Side note, I didn't remember that option existing but apparently I wrote it.

File: components/mpas-albany-landice/src/mode_forward/mpas_li_advection.F [hunk 2]
```diff
@@ -1301,7 +1294,6 @@ subroutine tracer_setup(&
       !       May need to increase maxTracers in the Registry.
       if ( (trim(config_tracer_advection) == 'fct') .or. &
            (trim(config_thickness_advection) == 'fct') ) then
-         call li_tracer_advection_fct_shared_init(geometryPool, err1)
          call li_tracer_advection_fct_init(err2)
 
          if (err1 /= 0 .or. err2 /= 0) then
```
Comment (matthewhoffman) [inline components/mpas-albany-landice/src/mode_forward/mpas_li_advection.F:1299]:
```suggestion
         if (err2 /= 0) then
```

Might be able to remove `err2` from the subroutine and replace this instance with `err`.  I haven't looked carefully.

**Review by matthewhoffman — CHANGES_REQUESTED** [review]
@trhille , sorry for letting this sit for so long.  After spending some time with it, I'm confident the implementation is correct, and the reason for keeping the FCT masks fixed between RK stages makes sense.  The only question left is what additional testing, if any, should be done.  I can see two levels of complexity here: 1. ensure that for changing ice extent (retreat or advance) the code runs without error and produces results that look as expected, or 2. for changing ice extent, evaluate order of accuracy with and without this change.  We know we can only expect first order accuracy when the margin changes, but perhaps before this change we are seeing noticeably worse than first order.  I'm content to go with option 1.  What do you think about testing?

Other than that, I added two very minor requests.

File: components/mpas-albany-landice/src/mode_forward/mpas_li_time_integration_fe_rk.F (hunk 1 above)
Comment (trhille) [reply components/mpas-albany-landice/src/mode_forward/mpas_li_time_integration_fe_rk.F:292]: Probably a copy-paste oversight. I'll remove.

File: components/mpas-albany-landice/src/mode_forward/mpas_li_advection.F (hunk 2 above)
Comment (trhille) [reply components/mpas-albany-landice/src/mode_forward/mpas_li_advection.F]: Fixed in https://github.com/MALI-Dev/E3SM/pull/110/commits/c0f079e10a3c5fc33034db3a095a3f8c873b70bd.

File: components/mpas-albany-landice/src/mode_forward/mpas_li_time_integration_fe_rk.F (hunk 1 above)
Comment (trhille) [reply components/mpas-albany-landice/src/mode_forward/mpas_li_time_integration_fe_rk.F]: Removed in https://github.com/MALI-Dev/E3SM/pull/110/commits/c0f079e10a3c5fc33034db3a095a3f8c873b70bd.

**Comment by trhille** [comment]
@matthewhoffman, regarding testing, we know this is a necessary change, but it's probably not sufficient. So I think option 1 (ensuring reasonable behavior without formally establishing convergence) is fine. 
@mperego, what do you think?

**Comment by mperego** [comment]
> @matthewhoffman, regarding testing, we know this is a necessary change, but it's probably not sufficient. So I think option 1 (ensuring reasonable behavior without formally establishing convergence) is fine. @mperego, what do you think?

I agree.

**Comment by trhille** [comment]
Testing
-------
I tested this on the 2km mismip mesh with a two-phase experiment with this branch rebased onto `MALI-Dev/develop`, compared with the head of `MALI-Dev/develop`. I used 3-stage RK3, 3rd order advection with `config_advection_coef_3rd_order = 1.0`. Basal melting and thermodynamics were turned off. Advective and calvign CFL fractions were set to 0.8.

Phase 1: 50-yr retreat with a specified calving velocity of 2km/yr. Flow velocity is ~1km/yr so this leads to ~50 km retreat.
Phase 2: 30-yr advance with calving turned off. I ran this for a shorter duration so that the ice shelf would not reach the calving mask at the domain edge.

Phase 1 results:
Top panel: using head of `develop` at end of phase 1
Middle panel: this branch rebased onto `develop`at end of phase 1
bottom: difference
image

**Comment by matthewhoffman** [comment]
@trhille thanks for adding that test results, including the thickness differences.  The differences are small enough that I think proceeding with this PR makes sense.  It looks like the old method retreats about one row farther and advances about row cell farther than the new method, but I can't think of an argument if which is more correct either way.  

(Side note - both versions seem to have a slight elevated ridge along the calving front in the advancing case, which is not realistic.  That's unrelated to this PR of course, but interesting to note.)

@mperego , I know you already approved this ages ago, but based on these new results, do you think we should look further into anything?  If not, I will go ahead and approve and merge as-is.

**Comment by mperego** [comment]
Yes, I think it's fine to merge.

**Comment by trhille** [comment]
Note: I've confirmed that the clear thickness artifacts at the advancing ice front in the figures above are the result of using too large a time step. A run with an advective CFL fraction of 0.2 exhibits a much smoother ice front. here's a transect from that run in blue vs from the CFL fraciton 0.8 run in the figures above, after fifteen years of advance:
image

**Comment by trhille** [comment]
I've confirmed that this change leads to minimal differences in the Halfar dome convergence test, using 2nd order FCT and RK2. One panel is head of `develop`, the other is this branch rebased onto `develop`:
image

Here is the thickness difference at year 200 (calculated using `ncdiff`, so it says `year = 0.00` because it differenced `daysSinceStart` as well as the relevant fields):
image

**Review by matthewhoffman — APPROVED** [review]
@trhille , thanks for making those small changes and doing all that additional testing.  It is quite informative to know how this change affects the convergence tests, and I feel more confident proceeding knowing the differences are so slight.

I reran the full_integration suite on perlmutter and everything passes.

**Pull request merged**